src/System.Linq.Parallel/tests/DegreeOfParallelismTests.cs

```diff
@@ -1,140 +1,86 @@
 // Copyright (c) Microsoft. All rights reserved.
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
-using Xunit;
 using System;
 using System.Collections.Generic;
 using System.Linq;
-using System.Text;
 using System.Threading;
 using System.Threading.Tasks;
+using Xunit;
 
 namespace Test
 {
     public class DegreeOfParallelismTests
     {
-        [Fact]
-        [OuterLoop]
-        public static void RunDopTests()
+        public static IEnumerable<object[]> DegreeData(object[] counts, object[] degrees)
         {
-            RunDopTest(1, false);
-            RunDopTest(4, false);
-            RunDopTest(512, false);
-            RunDopTest(513, true);
+            foreach (object[] results in UnorderedSources.Ranges(counts.Cast<int>()))
+            {
+                foreach (int degree in degrees.DefaultIfEmpty(results[1]).Cast<int>())
+                {
+                    yield return new object[] { results[0], results[1], degree };
+                }
+            }
         }
 
-        [Fact]
+        [Theory]
+        [MemberData("DegreeData", new[] { 1024 }, new[] { 1, 4, 512 })]
         [OuterLoop]
-        public static void RunDopBarrierTest()
+        public static void DegreeOfParallelism(Labelled<ParallelQuery<int>> labelled, int count, int degree)
         {
-            RunDopBarrierTestCore(1);
-            RunDopBarrierTestCore(4);
-            RunDopBarrierTestCore(32);
-            // These variations take way too long and hit the 100s timeout.
-            //RunDopBarrierTest(64);
-            //RunDopBarrierTest(512);
+            Assert.Equal(Functions.SumRange(0, count), labelled.Item.WithDegreeOfParallelism(degree).Sum());
         }
 
-        [Fact]
+        [Theory]
+        [MemberData("DegreeData", new[] { 1, 4, 32 }, new object[] { })]
         [OuterLoop]
-        public static void RunDopPipeliningTest()
+        public static void DegreeOfParallelism_Barrier(Labelled<ParallelQuery<int>> labelled, int count, int degree)
         {
-            RunDopPipeliningTestCore(1000000, 1);
-            RunDopPipeliningTestCore(1000000, 4);
-            RunDopPipeliningTestCore(1000000, 64);
-            RunDopPipeliningTestCore(1000000, 512);
+            //int minWorker, minAsyncIO;
+            //ThreadPool.GetMinThreads(out minWorker, out minAsyncIO);
+            //ThreadPool.SetMinThreads(degree, minAsyncIO);
```

Comment:
We had to remove these because they are not portable to other thread pool implementations; not every thread pool gives this kind of control.  I believe the issue here is limited to test scenarios only; production code should not be making assumptions about thread-injection policy.  Given that, I could see value in adding some sort of internal contract to control the number of threads in the thread pool, which we only support on some platforms, and perhaps the test could fall back to some other behavior on platforms that don’t support this.  However, this is certainly not something we support now.

Eric

From: Stephen Toub [mailto:notifications@github.com]
Sent: Tuesday, June 2, 2015 7:46 AM
To: dotnet/corefx
Cc: Eric Eilebrecht
Subject: Re: [corefx] System.Linq.Parallel Test - initial cleanup (#1599)


In src/System.Linq.Parallel/tests/DegreeOfParallelismTests.cs<https://github.com/dotnet/corefx/pull/1599#discussion_r31530064>:

>          {

> -            RunDopPipeliningTestCore(1000000, 1);

> -            RunDopPipeliningTestCore(1000000, 4);

> -            RunDopPipeliningTestCore(1000000, 64);

> -            RunDopPipeliningTestCore(1000000, 512);

> +            //int minWorker, minAsyncIO;

> +            //ThreadPool.GetMinThreads(out minWorker, out minAsyncIO);

> +            //ThreadPool.SetMinThreads(degree, minAsyncIO);

Hmm, if it's not exposed in the contract, @ericeil<https://github.com/ericeil> / @weshaggard<https://github.com/weshaggard>, what's the recommended way for folks to handle situations like this, where concurrent work really needs a particular minimum number of threads to be available in the pool?

—
Reply to this email directly or view it on GitHub<https://github.com/dotnet/corefx/pull/1599/files#r31530064>.